### PR TITLE
Generic routing module (routing policies)

### PR DESCRIPTION
**Pull request by ipspace**
The generic routing module implements route maps (routing.policy) which can be used in bgp.policy plugin with bgp.policy.in and bgp.policy.out interface attributes.

bgp.policy plugin has been modified to use the shared route-map infrastructure and activate 'routing' module whenever
necessary.

The configuration templates for all devices supported by the bgp.policy plugin have been split into route-map and BGP parts, and the route-map part moved into the 'routing' module.

### CODE REVIEW

**Comment by ipspace** [comment]
Missing:

* bgp.policy should copy attribute definitions and device support data from the routing module
* routing module should check the feature support when a route-map is applied to a node
* test- and error cases
* documentation

Also:

* No work has been done (yet) on the bgp.policy attribute

**Comment by ipspace** [comment]
Implemented:

* Route map merging
* Device feature support
* bgp.policy.in and bgp.policy.out interface attributes
* Integration tests mirroring the bgp.policy integration tests
* Test and error cases for routing.policy

Missing:

* [ ] routing.policy data type errors stress test
* [ ] bgp.policy / routing.policy integration error tests
* [ ] documentation
* [ ] bgp.policy should copy attribute definitions and device support data from the routing module

**Comment by ipspace** [comment]
FWIW, this is ready for review. Will implement 'copy attributes from routing to bgp.policy' after the merge.

Also, I will probably submit the changes as several commits, not as a single PR.

File: tests/errors/bgp.policy-mix.yml.log
Comment (ssasso) [inline tests/errors/bgp.policy-mix.yml.log:1]: I guess this is not the real test result ;)

Comment (ipspace) [reply tests/errors/bgp.policy-mix.yml.log:1]: Weird. Will figure it out. Thank you!

**Review by ssasso — APPROVED** [review]
All looks wonderful ;) thank you!
(also for the "merge" of arubacx route maps with ios)

(I also did some additional tests (with arubacx and vyos), and all worked)

**Comment by ipspace** [comment]
> All looks wonderful ;) thank you! (also for the "merge" of arubacx route maps with ios)
> 
> (I also did some additional tests (with arubacx and vyos), and all worked)

Thanks a million ;) I know it was quite a lot to go through.

**Comment by ipspace** [comment]
Implemented in #1238 and #1239

**Comment by ipspace** [comment]
The remaining work has been copied into #1240

**Pull request closed without merging**